File: app/src/main/java/org/sorz/lab/tinykeepass/BaseActivity.java
package org.sorz.lab.tinykeepass;

import android.app.Activity;
import android.app.KeyguardManager;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.security.keystore.UserNotAuthenticatedException;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import org.sorz.lab.tinykeepass.auth.FingerprintDialogFragment;
import org.sorz.lab.tinykeepass.auth.SecureStringStorage;
import org.sorz.lab.tinykeepass.keepass.OpenKeePassTask;

import java.security.KeyException;
import java.util.List;
import java.util.function.Consumer;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;

import de.slackspace.openkeepass.domain.KeePassFile;

import static org.sorz.lab.tinykeepass.DatabaseSetupActivity.AUTH_METHOD_FINGERPRINT;
import static org.sorz.lab.tinykeepass.DatabaseSetupActivity.AUTH_METHOD_NONE;
import static org.sorz.lab.tinykeepass.DatabaseSetupActivity.AUTH_METHOD_SCREEN_LOCK;
import static org.sorz.lab.tinykeepass.DatabaseSetupActivity.AUTH_METHOD_UNDEFINED;
import static org.sorz.lab.tinykeepass.DatabaseSetupActivity.PREF_KEY_AUTH_METHOD;


public abstract class BaseActivity extends AppCompatActivity
        implements FingerprintDialogFragment.OnFragmentInteractionListener {
    private final static String TAG = MainActivity.class.getName();
    private final static int REQUEST_CONFIRM_DEVICE_CREDENTIAL = 100;
    private final static int REQUEST_SETUP_DATABASE = 101;

    private SharedPreferences preferences;
    private KeyguardManager keyguardManager;
    private SecureStringStorage secureStringStorage;
    private Consumer<List<String>> onKeyRetrieved;
    private Consumer<String> onKeyAuthFailed;
    private Runnable onKeySaved;
    private List<String> keysToEncrypt;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        preferences = PreferenceManager.getDefaultSharedPreferences(this);
        keyguardManager = (KeyguardManager) getSystemService(KEYGUARD_SERVICE);
        try {
            secureStringStorage = new SecureStringStorage(this);
        } catch (SecureStringStorage.SystemException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case REQUEST_CONFIRM_DEVICE_CREDENTIAL:
                if (resultCode == RESULT_OK)
                    getKey();
                else
                    authFail(getString(R.string.fail_to_auth));
                break;
            case REQUEST_SETUP_DATABASE:
                if (resultCode == RESULT_OK)
                    getKey();
                else
                    authFail(getString(R.string.fail_to_decrypt));
                break;
            default:
                super.onActivityResult(requestCode, resultCode, data);
        }
    }

    private void authFail(String message) {
        if (onKeyAuthFailed != null)
            onKeyAuthFailed.accept(message);
        onKeyAuthFailed = null;
        onKeySaved = null;
        onKeyRetrieved = null;
        keysToEncrypt = null;
    }

    @Override
    public void onFingerprintCancel() {
        authFail(getString(R.string.fail_to_auth));
    }

    @Override
    public void onFingerprintSuccess(Cipher cipher) {
        if (onKeyRetrieved != null)
            decryptKey(cipher);
        else if (onKeySaved != null)
            encryptKeys(cipher);
    }

    protected void getDatabaseKeys(Consumer<List<String>> onKeyRetrieved,
                                   Consumer<String> onKeyAuthFailed) {
        this.onKeyRetrieved = onKeyRetrieved;
        this.onKeyAuthFailed = onKeyAuthFailed;
        getKey();
    }

    protected void saveDatabaseKeys(@NonNull List<String> keys, Runnable onKeySaved,
                                    Consumer<String> onKeyAuthFailed) {
        this.onKeySaved = onKeySaved;
        this.onKeyAuthFailed = onKeyAuthFailed;
        saveKey(keys);
    }

    @Override
    public void onKeyException(KeyException e) {
        // Key is invalided, have to reconfigure passwords.
        Intent intent = new Intent(this, DatabaseSetupActivity.class);
        startActivityForResult(intent, REQUEST_SETUP_DATABASE);
    }

    /**
     * Authenticate user, then call {@link #encryptKeys(Cipher)} to save keys.
     * Finally, either {@link #onKeySaved} or {@link #onKeyAuthFailed} will be called.
     */
    private void saveKey(@NonNull List<String> keys) {
        keysToEncrypt = keys;
        int authMethod = preferences.getInt(PREF_KEY_AUTH_METHOD, AUTH_METHOD_UNDEFINED);
        try {
            switch (authMethod) {
                case AUTH_METHOD_NONE:
                    secureStringStorage.generateNewKey(false, -1);
                    encryptKeys(null);
                    break;
                case AUTH_METHOD_SCREEN_LOCK:
                    secureStringStorage.generateNewKey(true, 60);
                    Intent intent = keyguardManager.createConfirmDeviceCredentialIntent(
                            getString(R.string.auth_key_title),
                            getString(R.string.auth_key_description));
                    startActivityForResult(intent, REQUEST_CONFIRM_DEVICE_CREDENTIAL);
                    break;
                case AUTH_METHOD_FINGERPRINT:
                    secureStringStorage.generateNewKey(true, -1);
                    getFragmentManager().beginTransaction()
                            .add(FingerprintDialogFragment.newInstance(Cipher.ENCRYPT_MODE),
                                "fingerprint")
                            .commit();
                    break;
            }
        } catch (SecureStringStorage.SystemException e) {
            throw new RuntimeException("cannot generate new key", e);
        }
    }

    private void encryptKeys(Cipher cipher) {
        try {
            if (cipher == null)
                cipher = secureStringStorage.getEncryptCipher();
            secureStringStorage.put(cipher, keysToEncrypt);
        } catch (UserNotAuthenticatedException e) {
            Log.e(TAG, "cannot get cipher from system", e);
            onKeyAuthFailed.accept("cannot get cipher from system");
            return;
        } catch (SecureStringStorage.SystemException | KeyException e) {
            throw new RuntimeException("cannot get save keys", e);
        }
        onKeySaved.run();
        onKeySaved = null;
        onKeyAuthFailed = null;
        keysToEncrypt = null;
    }

    /**
     * Authenticate user, then call {@link #decryptKey(Cipher)} to get keys.
     * Finally, either {@link #onKeyRetrieved} or {@link #onKeyAuthFailed} will be called.
     */
    private void getKey() {
        int authMethod = preferences.getInt(PREF_KEY_AUTH_METHOD, AUTH_METHOD_UNDEFINED);
        switch (authMethod) {
            case AUTH_METHOD_UNDEFINED:
                authFail(getString(R.string.broken_keys));
                break;
            case AUTH_METHOD_NONE:
            case AUTH_METHOD_SCREEN_LOCK:
                try {
                    Cipher cipher = secureStringStorage.getDecryptCipher();
                    decryptKey(cipher);
                } catch (UserNotAuthenticatedException e) {
                    // should do authentication
                    Intent intent = keyguardManager.createConfirmDeviceCredentialIntent(
                            getString(R.string.auth_key_title),
                            getString(R.string.auth_key_description));
                    startActivityForResult(intent, REQUEST_CONFIRM_DEVICE_CREDENTIAL);
                } catch (KeyException e) {
                    onKeyException(e);
                } catch (SecureStringStorage.SystemException e) {
                    throw new RuntimeException(e);
                }
                break;
            case AUTH_METHOD_FINGERPRINT:
                getFragmentManager().beginTransaction()
                        .add(FingerprintDialogFragment.newInstance(Cipher.DECRYPT_MODE),
                                "fingerprint")
                        .commit();
                break;
        }
    }

    private void decryptKey(Cipher cipher) {
        List<String> keys;
        try {
            keys = secureStringStorage.get(cipher);
        } catch (BadPaddingException | IllegalBlockSizeException e) {
            Log.w(TAG, "fail to decrypt keys", e);
            authFail(getString(R.string.fail_to_decrypt));
            return;
        }
        if (keys == null || keys.size() < 2) {
            authFail(getString(R.string.broken_keys));
            return;
        }
        onKeyRetrieved.accept(keys);
        onKeyAuthFailed = null;
        onKeyRetrieved = null;
    }

    protected void openDatabase(String masterKey, Consumer<KeePassFile> onSuccess) {
        new OpenTask(this, masterKey, onSuccess).execute();
    }

    protected SecureStringStorage getSecureStringStorage() {
        return secureStringStorage;
    }

    protected SharedPreferences getPreferences() {
        return preferences;
    }

    private static class OpenTask extends OpenKeePassTask {
        private final Consumer<KeePassFile> onSuccess;

        OpenTask(Activity activity, String masterKey, Consumer<KeePassFile> onSuccess) {
            super(activity, masterKey);
            // TODO: memory leaks?
            this.onSuccess = onSuccess;
        }

        @Override
        protected void onPostExecute(KeePassFile db) {
            super.onPostExecute(db);
            if (db != null)
                onSuccess.accept(db);
        }
    }
}

File: app/src/main/java/org/sorz/lab/tinykeepass/BaseActivity.kt
package org.sorz.lab.tinykeepass

import android.content.Intent
import android.content.SharedPreferences
import android.os.*
import android.security.keystore.UserNotAuthenticatedException
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.biometric.BiometricManager.Authenticators.*
import androidx.biometric.BiometricPrompt
import androidx.biometric.BiometricPrompt.ERROR_HW_NOT_PRESENT
import androidx.biometric.BiometricPrompt.ERROR_NO_DEVICE_CREDENTIAL
import androidx.preference.PreferenceManager

import org.sorz.lab.tinykeepass.auth.SecureStringStorage

import javax.crypto.Cipher
import java.security.KeyException
import java.lang.Exception

import kotlinx.coroutines.suspendCancellableCoroutine
import org.sorz.lab.tinykeepass.keepass.OpenDatabaseError
import org.sorz.lab.tinykeepass.keepass.openKeePassTask
import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException


private val TAG = MainActivity::class.java.name

abstract class BaseActivity : AppCompatActivity() {
    protected val preferences: SharedPreferences by lazy(LazyThreadSafetyMode.NONE) {
        PreferenceManager.getDefaultSharedPreferences(this)
    }
    private lateinit var biometricPrompt: BiometricPrompt
    protected val secureStringStorage by lazy(LazyThreadSafetyMode.NONE) {
        try {
            SecureStringStorage(this)
        } catch (e: SecureStringStorage.SystemException) {
            throw RuntimeException(e)
        }
    }
    private val setupDatabase = registerForActivityResult(StartActivityForResult()) {
        // if (it.resultCode == RESULT_OK) getKey()
        // TODO
    }
    private var biometricAuthResult: Continuation<Cipher>? = null


    private fun launchSetupDatabase() {
        val intent = Intent(this@BaseActivity, DatabaseSetupActivity::class.java)
        setupDatabase.launch(intent)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        biometricPrompt = BiometricPrompt(this, object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                biometricAuthResult?.resume(result.cryptoObject!!.cipher!!)
            }
            override fun onAuthenticationError(code: Int, msg: CharSequence) {
                biometricAuthResult?.resumeWithException(AuthKeyError(msg.toString()))
                if (code == ERROR_NO_DEVICE_CREDENTIAL || code == ERROR_HW_NOT_PRESENT) {
                    // Reconfigure passwords
                    launchSetupDatabase()
                }
            }
        })
    }

    protected suspend fun getDatabaseKeys(): List<String> {
        val decryptCipher = secureStringStorage.decryptCipher
        val cipher = when (preferences.getInt(PREF_KEY_AUTH_METHOD, AUTH_METHOD_UNDEFINED)) {
            AUTH_METHOD_NONE, AUTH_METHOD_SCREEN_LOCK ->
                try {
                    return decryptKey(decryptCipher)
                } catch (e: UserNotAuthenticatedException) {
                    // should do authentication
                    authenticateCipher(decryptCipher, true)
                } catch (e: KeyException) {
                    // Reconfigure passwords
                    launchSetupDatabase()
                    throw AuthKeyError(getString(R.string.broken_keys))
                } catch (e: SecureStringStorage.SystemException) {
                    throw RuntimeException(e)
                }
            AUTH_METHOD_FINGERPRINT -> authenticateCipher(decryptCipher, false)
            else -> throw AuthKeyError(getString(R.string.broken_keys))
        }
        return decryptKey(cipher)
    }

    private suspend fun authenticateCipher(cipher: Cipher, useDeviceCredential: Boolean): Cipher {
        val prompt = BiometricPrompt.PromptInfo.Builder()
                .setTitle(getString(R.string.auth_key_title))
                .setDescription(getString(R.string.auth_key_description))
                .setNegativeButtonText(getText(android.R.string.cancel))
                .setConfirmationRequired(false)
                .setAllowedAuthenticators(
                    if (useDeviceCredential) BIOMETRIC_WEAK or DEVICE_CREDENTIAL
                    else BIOMETRIC_STRONG
                )
                .build()
        val crypto = BiometricPrompt.CryptoObject(cipher)
        return suspendCancellableCoroutine { cont ->
            biometricAuthResult = cont
            biometricPrompt.authenticate(prompt, crypto)
        }
    }

    suspend fun saveDatabaseKeys(keys: List<String>) {
        val encryptCipher = secureStringStorage.encryptCipher
        val cipher = when (preferences.getInt(PREF_KEY_AUTH_METHOD, AUTH_METHOD_UNDEFINED)) {
            AUTH_METHOD_NONE -> {
                secureStringStorage.generateNewKey(false, -1)
                return encryptKeys(encryptCipher, keys)
            }
            AUTH_METHOD_SCREEN_LOCK -> {
                secureStringStorage.generateNewKey(true, 60)
                authenticateCipher(encryptCipher, true)
            }
            AUTH_METHOD_FINGERPRINT -> {
                secureStringStorage.generateNewKey(true, -1)
                authenticateCipher(encryptCipher, false)
            }
            else -> throw AuthKeyError(getString(R.string.broken_keys))
        }
        encryptKeys(cipher, keys)
    }

    private fun encryptKeys(cipher: Cipher, keys: List<String>) {
        try {
            secureStringStorage.put(cipher, keys)
        } catch (e: UserNotAuthenticatedException) {
            Log.e(TAG, "cannot get cipher from system", e)
            throw AuthKeyError("cannot get cipher from system")
        }
    }

    private fun decryptKey(cipher: Cipher): List<String> {
        val keys = try {
            secureStringStorage.get(cipher)
        } catch (e: Exception) {
            Log.w(TAG, "fail to decrypt keys", e)
            throw AuthKeyError(getString(R.string.fail_to_decrypt))
        }
        if (keys == null || keys.size < 2)
            throw AuthKeyError(getString(R.string.broken_keys))
        return keys
    }

    @Throws(AuthKeyError::class, OpenDatabaseError::class)
    protected suspend fun openDatabase() {
        val keys = getDatabaseKeys()
        openKeePassTask(this, keys[0])
    }
}

open class AuthKeyError(message: String) : Exception(message)
